src/resolve/resolved-manager.c

```diff
@@ -442,28 +442,40 @@ static int determine_hostname(char **full_hostname, char **llmnr_hostname, char
         return 0;
 }
 
-static const char *fallback_hostname(void) {
+static char* fallback_hostname(void) {
 
-        /* Determine the fall back hostname. For exposing this system to the outside world, we cannot have it to be
-         * "localhost" even if that's the compiled in hostname. In this case, let's revert to "linux" instead. */
+        /* Determine the fall back hostname. For exposing this system to the outside world, we cannot have it
+         * to be "localhost" even if that's the default hostname. In this case, let's revert to "linux"
+         * instead. */
 
-        if (is_localhost(FALLBACK_HOSTNAME))
-                return "linux";
+        _cleanup_free_ char *n = get_default_hostname();
+        if (!n)
+                return NULL;
+
+        if (is_localhost(n))
+                return strdup("linux");
 
-        return FALLBACK_HOSTNAME;
+        return TAKE_PTR(n);
 }
 
 static int make_fallback_hostnames(char **full_hostname, char **llmnr_hostname, char **mdns_hostname) {
-        _cleanup_free_ char *n = NULL, *m = NULL;
-        char label[DNS_LABEL_MAX], *h;
+        _cleanup_free_ char *_free_hn = NULL, *n = NULL, *m = NULL, *h = NULL;
+        char label[DNS_LABEL_MAX];
         const char *p;
         int r;
 
         assert(full_hostname);
         assert(llmnr_hostname);
         assert(mdns_hostname);
 
-        p = fallback_hostname();
+        p = _free_hn = fallback_hostname();
+        if (!p)
+                return log_oom();
+
+        h = strdup(p);
```

Comment:
you can just use _free_hn here? i.e. rename _free_hn → h? and drop `p`?

Reply:
`p` is still needed, because dns_label_unescape moves the pointer. But yeah, _free_hn can be dropped.